Ryujinx.Graphics.Gpu/Memory/Buffer.cs

```diff
@@ -194,341 +154,446 @@ public bool FullyContains(ulong address, ulong size)
+
+            for (int index = 0; index < _regions.Length; index++)
+            {
+                BufferRegion region = _regions[index];
+
+                handles[index] = new RegionHandleSegment(baseOffset + region.BaseOffset, region.Size, region.GetTrackingHandles());
+            }
+
+            return handles;
         }
 
         /// <summary>
-        /// Flush modified ranges of the buffer from another thread.
-        /// This will flush all modifications made before the active SyncNumber was set, and may block to wait for GPU sync.
+        /// Gets the tracking handles on a given sub-range of this buffer.
         /// </summary>
-        /// <param name="address">Address of the memory action</param>
-        /// <param name="size">Size in bytes</param>
-        public void ExternalFlush(ulong address, ulong size)
+        /// <param name="baseOffset">Offset where the handles will be placed on the new buffer that will inherit them</param>
+        /// <param name="bufferOffset">Start offset of the range to get the tracking handles from</param>
+        /// <param name="size">Size in bytes of the range</param>
+        /// <returns>The tracking handles at the specified range</returns>
+        public IEnumerable<RegionHandleSegment> GetTrackingHandlesSlice(ulong baseOffset, ulong bufferOffset, ulong size)
         {
-            _context.Renderer.BackgroundContextAction(() =>
+            ulong bufferEndOffset = bufferOffset + size;
+            int index = FindStartIndex(bufferOffset);
+
+            List<RegionHandleSegment> handles = new List<RegionHandleSegment>();
+
+            for (; index < _regions.Length; index++)
             {
-                var ranges = _modifiedRanges;
+                BufferRegion region = _regions[index];
+
+                if (region.BaseOffset >= bufferEndOffset)
+                {
+                    break;
+                }
+
+                ulong regionEndOffset = region.BaseOffset + region.Size;
+
+                ulong clampedOffset = Math.Max(bufferOffset, region.BaseOffset);
+                ulong clampedEndOffset = Math.Min(bufferEndOffset, regionEndOffset);
+                ulong clampedSize = clampedEndOffset - clampedOffset;
 
-                if (ranges != null)
+                IEnumerable<IRegionHandle> regionHandles = region.GetTrackingHandles();
+
+                if (clampedOffset > region.BaseOffset || clampedEndOffset < regionEndOffset)
                 {
-                    (address, size) = PageAlign(address, size);
-                    ranges.WaitForAndGetRanges(address, size, Flush);
+                    ulong skipSize = clampedOffset - region.BaseOffset;
+
+                    int skipCount = (int)(skipSize / BufferRegion.GranularBufferThreshold);
+                    int takeCount = (int)(clampedSize / BufferRegion.GranularBufferThreshold);
+
+                    regionHandles = regionHandles.Skip(skipCount).Take(takeCount);
```

Comment:
```suggestion
                    regionHandles = regionHandles.Take(skipCount..skipCount + takeCount);
```
Take() accepts Range since .NET 6, avoids some allocations compared to Skip().Take()